### ISSUE TITLE
Composite predicates

### ISSUE THREAD

**Issue by dlubarov** [mir-core/mips#7]
In Zexe, it seems natural for applications to require that records have a certain standardized birth predicate, while not imposing any requirement on death predicates. An example is the token scheme described in the Zexe paper, where each token record is required to have a particular birth predicate `Φ_b` which enforces balance conservation. A token record can have any death predicate, so custom death predicates can be written to verify arbitrary signatures, enforce time locks, and so forth.

However, a problem arises when an application needs to standardize death predicates. For example, suppose we were implementing a DEX. The natural approach would be to write a death predicate `Φ_d`, declaring "any transaction may consume this record if it also creates X units of token Y, spendable by Z". To submit an order, a user would move some tokens into a record with `Φ_d`, then register it in some decentralized order book.

But now that we've mandated that every token order record use both `Φ_b` and `Φ_d`, there is no room left for customization. `Φ_d` could support some standardized signature scheme, but users would lose the flexibility to use arbitrary schemes. If there were two competing DEXs using different death predicates, users would need to pick one or the other; it wouldn't be possible to offer the same tokens on two different DEXs.

Maybe these limitation wouldn't be a big deal in practice, but let's explore some possible solutions.

## Composite predicates

In the example above, what we want is multiple death predicates with a disjunctive relationship. Then the user could create a record with a death predicate for DEX A, another for DEX B, and another for an owner-authenticated transfer (in case the order doesn't get filled).

However, disjunction alone does not seem sufficient. Suppose that a DEX was being overloaded with short-lived orders from market makers. To address the problem, they might want to require that orders remain in the order book for a certain time period before the funds can be transferred out. To meet this requirement, order records would need to have a death predicate structure like

    order is filled OR (owner authorizes a transfer AND time lock has expired)

## Composition via circuits

A straightforward way to support this sort of composition would be to give each record two small logical circuits, one for birth and one for death. Each wire would be either a constant or a predicate. These circuits would be private by default; users could prove their satisfiability in zero knowledge after first proving certain predicates to be satisfiable.

## Composition via plugins

Another approach would be a plugin model, where a predicate can designate certain wires in its circuit as hooks, and arbitrary subpredicates can be attached to those hooks.

This model would be more powerful in a sense, since the top-level predicate would be a full-fledged Zexe predicate with access to all the transaction data. It could require certain subpredicates only in certain contexts, depending on what the transaction was doing. For example, imagine an application where users can decide whether or not to publicly log each operation. The top-level predicate could check if there was a data command (see #3) writing to the log structure, and if so, require a subpredicate which validates the content of the log.

I don't think this extra power is really necessary. In our example, the log validation subpredicate itself could check the list of data commands, and output 1 for a transaction with no logging. However, having this logic in the top-level predicate could be more performant, as users would not need to generate an argument for the log validation subpredicate if they knew that the top-level predicate would not require it.

Furthermore, this approach would make it possible to replace birth and death predicates with a single predicate per record. The top-level predicate could check whether the record it cares about was being created or consumed, and potentially require different subpredicates accordingly. That said, there are some tradeoffs here and I'm not sure whether it would be a good idea overall; need to give this more thought.

Going back to our token example, a token record with a single predicate could look something like this. This is for a record holding 100 DAI, whose owner was offering it on a DEX in exchange for 10 MKR.

    StandardToken(
        token = "DAI",
        units = 100,
        transfer_authorization = DexOrderFilled(
            ask_token = "MKR",
            ask_units = 10,
            submitted_at = "2018-12-03",
            owner_authorization = Sha3Signature(
                pk = <owner's public key>)))

Here `DexOrderFilled` would allow another DEX user to claim the 100 DAI if they simultaneously transferred 10 MKR to the owner. `owner_authorization` would also be used to let the owner cancel their order, perhaps after a time lock expired.

## Open questions

Is the Zexe limitation significant enough to warrant nontrivial solutions like these? If so, which model would work best, circuits or plugins? If we go with the plugin model, should be keep separate birth and death predicates, or combine them?

Should we generate a separate argument for each subpredicate, then compose them with recursive arguments? Or should we generate one big composite circuit, then generate a single argument for the whole thing? The answer might depend on which argument system we use.

**Comment by bpfarmer** [comment]
> A straightforward way to support this sort of composition would be to give each record two small logical circuits, one for birth and one for death. Each wire would be either a constant or a predicate. These circuits would be private by default; users could prove their satisfiability in zero knowledge after first proving certain predicates to be satisfiable.

This makes sense. Just so I understand properly, if I had three predicates `p_1, p_2, p_3`, and the death predicate would be a circuit encoding `p1 AND (p_2 OR p_3)`, I would provide a succinct proof that an input satisfies `p_1, p_2` that would be an input to a proof showing that the death predicate circuit was satisfied?

> Furthermore, this approach would make it possible to replace birth and death predicates with a single predicate per record. The top-level predicate could check whether the record it cares about was being created or consumed, and potentially require different subpredicates accordingly.

Hmm I might be misunderstanding you, but wouldn't this not work, since certain records will need to require strict equality in birth or death predicates? I'm thinking of the basic example of a token where each birth predicate requires conservation of balances and an identical birth predicate in the consumed record. 

> Should we generate a separate argument for each subpredicate, then compose them with recursive arguments? Or should we generate one big composite circuit, then generate a single argument for the whole thing? The answer might depend on which argument system we use.

Would there be an efficiency benefit in using a sort of Merkleized Abstract Syntax Tree? I just mean a binary tree where each non-leaf node is designated either an AND or an OR node, and then each leaf is a predicate? I'm not sure if it would be easier or more efficient to prove satisfiability in this context, since you could ignore predicates that you didn't need? I'm still developing an intuition for SNARKs.

**Comment by bpfarmer** [comment]
I think that the MAST solution might be worth thinking about further, depending on the proving time

**Comment by dlubarov** [comment]
> This makes sense. Just so I understand properly, if I had three predicates `p_1, p_2, p_3`, and the death predicate would be a circuit encoding `p1 AND (p_2 OR p_3)`, I would provide a succinct proof that an input satisfies `p_1, p_2` that would be an input to a proof showing that the death predicate circuit was satisfied?

Yeah exactly, that's what I was thinking. Or we could merge all the predicates into one circuit and generate a single proof; I think both approaches are fine so it's just a question of performance.

> Hmm I might be misunderstanding you, but wouldn't this not work, since certain records will need to require strict equality in birth or death predicates? I'm thinking of the basic example of a token where each birth predicate requires conservation of balances and an identical birth predicate in the consumed record. I suppose that it would be possible to force only monotonically-increasing constraints such that consuming records would include consumed records birth predicates, but this seems like it could be complicated.

I was thinking the token contract could have a plugin to control whether a transfer was authorized or not (`transfer_authorization` in the example). So `StandardToken` would have two checks

- Look at all `StandardToken` records in the transaction (ignoring their plugins), and check that balance is conserved.
- If the record associated with me is being consumed, then check that `transfer_authorization` is satisfied.

In that case the two checks directly correspond to birth and death predicates in the original Zexe model, so we wouldn't really gain anything. But I think there would be other use cases where having a single predicate is more natural, like a dapp with an admin role (or committee), where certain records can be created *or* consumed as long as the admin approves.

> Would there be an efficiency benefit in using a sort of Merkleized Abstract Syntax Tree? I just mean a binary tree where each non-leaf node is designated either an AND or an OR node, and then each leaf is a predicate? I'm not sure if it would be easier or more efficient to prove satisfiability in this context, since you could ignore predicates that you didn't need? I'm still developing an intuition for SNARKs.

Yeah I think that makes sense, except why a Merklized tree? It would definitely be nice to be able to skip proofs for certain subpredicates, like `p_3` in your example.

**Comment by dlubarov** [comment]
> I think that this could get a bit messy, since records would need to inspect the contents of a predicate to ensure that it was a valid StandardToken predicate with plugins, rather than just comparing the hash of a predicate verification key as in Zexe.

Yeah I see what you mean. I think it should be possible to inspect subpredicates (and ideally our tooling should make it easy), but with the birth/death paradigm, it should rarely be necessary.

> leaf nodes are hashed verification keys for each condition

Hm, would conditions correspond to predicates, or do you mean conditions in a different sense, like wire values? I don't quite understand. Taking your example of `p_1 AND (p_2 OR p_3)`, I was thinking the user would generate arguments for `p_1` and `p_2`, then an outer argument which recursively verifies all three (even though one will fail), then applies the two logic gates. Is your idea a way to improve upon that with an outer circuit which only verifies two arguments? That would be neat!

As an aside, I think the recursive approach could definitely speed up proof times in some cases. An extreme case would be a 1-of-100 multisig, where a merged circuit would be huge. With recursion we could generate just one signature verification argument, and then if we're clever, the outer argument could verify a single argument.

However, using predicate boundaries as our recursion boundaries might be far from optimal in some cases. For instance if we wrote a time lock predicate, that just involves an integer comparison, so the circuit would probably be much smaller than a recursive verification circuit. Plus if the time lock is required (i.e. not part of an OR), then every valid transaction will need to show that it's satisfied, so we wouldn't gain anything from recursion.

So I think if we want to really optimize proving times, we should let predicate authors (or compilers) decide where to set the recursion boundaries. Or if we wanted to keep it simple, we could just merge all predicates into one circuit, and that might be good enough in practice.

**Comment by dlubarov** [comment]
Ah I see, makes sense. It definitely seems like a useful optimization, though I think we could do it without a MAST, which might be slightly easier. The outer circuit could have one argument verifier for each OR gate, and the verification key could be a private input. Then for each OR gate, the outer circuit could check that the verification key was in a list, corresponding to the OR gate's operands. Finally, we wrap those two results in an AND gate, and that's the final output of the OR gate.

Does that make sense? I'm on mobile, but I can draw it out later if it's not clear.

I think it's almost equivalent to your idea? Just seems natural to me to generate a bit of circuitry per OR gate, so that our generated circuit would retain the structure of the original logical circuit.

**Comment by bpfarmer** [comment]
Hmm, maybe it would be good to draw it out, but yeah I think it's equivalent. In fact, it might be exactly equivalent depending on how the membership in the list of verification keys is checked - I assume that the accumulator would be a Merkle tree. I think the only question is which implementation is more efficient, and I have no idea. 

My only thought is that within the record, it might be difficult to close/open the commitment to the circuit that you describe - it might be hard to open the commitment, insert the appropriate verification key, regenerate the outer verification key, and then verify the proof. I sort of remember that there are some subtleties in the Zexe implementation because they use two different curves to prove different relations, and it can be tricky to share data between them - that's why they hash the verification keys w/ Blake2s, which is a relatively efficient implementation in both curves. Anyway, probably a bit premature to determine specific implementation details, but I think that the spirit of the approach seems to make a lot of sense.

**Comment by dlubarov** [comment]
Ah good point about the two curves. Yeah, let's revisit this later after deciding which argument system to use.
